Framework/API/inc/MantidAPI/MatrixWorkspace.h

```diff
@@ -151,6 +152,7 @@ class MANTID_API_DLL MatrixWorkspace : public IMDWorkspace,
   //----------------------------------------------------------------------
 
   /// Return the underlying ISpectrum ptr at the given workspace index.
+  /// Overriding classes must call invalidateCommonBinsFlag().
```

Comment:
`getSpectrum()` is the most often used entry point which gives potential write access to the X data. Unfortunately, it is declared pure virtual in `MatrixWorkspace` so we cannot really ensure that the common bins flag gets properly invalidated. The above comment doesn't really help. *All* deriving classes should invalidate the common bins flag in `getSpectrum()`.

For more safe invalidation, we'd need to implement `getSpectrum()` as a normal method which would do the invalidation and call a `protected` and `virtual` `getSpectrumImpl()` method which deriving classes could implement. Something like this:

```c++
class MatrixWorkspace {
public:
  ISpectrum &getSpectrum(const size_t index) const {
    invalidateCommonBinsFlag();
    return getSpectrumImpl(index);
  }
  // The const version can be left as-is.
  virtual const ISpectrum &getSpectrum(const size_t index) const = 0;
protected:
  // Deriving classes override this instead of getSpectrum above.
  virtual ISpectrum &getSpectrumImpl(const size_t index) const = 0;
}
```

Or that's how I see things now. Let me know there are better ways to do this or should I leave `getSpectrum()` as-is?